File: src/neetcode/Java/_47PermutationsII.java
package neetcode.Java;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class _47PermutationsII {
    List<List<Integer>> list = new ArrayList<>();
    int len;

    public List<List<Integer>> permuteUnique(int[] nums) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
        }
        len = nums.length;
        Helper(map, new ArrayList<>());
        return list;
    }

    public void Helper(HashMap<Integer, Integer> map, ArrayList<Integer> sub) {
        if (sub.size() == len) {
            list.add(new ArrayList<>(sub));
            return;
        }
        for (int i : map.keySet()) {
            if (map.get(i) > 0) {
                sub.add(i);
                map.put(i, map.get(i) - 1);
                Helper(map, sub);
                sub.remove(sub.size() - 1);
                map.put(i, map.get(i) + 1);
            }
        }
    }

    public static void main(String[] args) {
        _47PermutationsII obj = new _47PermutationsII();
        int[] nums = {1, 1, 2};
        System.out.println(obj.permuteUnique(nums));
    }
}

File: src/neetcode/kotlin/_47PermutationsII.kt
package neetcode.kotlin

class _47PermutationsII {
    var list: MutableList<List<Int>> = ArrayList()
    var len: Int = 0

    fun permuteUnique(nums: IntArray): List<List<Int>> {
        val map = HashMap<Int, Int>()
        for (i in nums.indices) {
            map[nums[i]] = map.getOrDefault(nums[i], 0) + 1
        }
        len = nums.size
        Helper(map, ArrayList())
        return list
    }

    fun Helper(map: HashMap<Int, Int>, sub: ArrayList<Int>) {
        if (sub.size == len) {
            list.add(ArrayList(sub))
            return
        }
        for (i in map.keys) {
            if (map[i]!! > 0) {
                sub.add(i)
                map[i] = map[i]!! - 1
                Helper(map, sub)
                sub.removeAt(sub.size - 1)
                map[i] = map[i]!! + 1
            }
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _47PermutationsII()
            val nums = intArrayOf(1, 1, 2)
            println(obj.permuteUnique(nums))
        }
    }
}
